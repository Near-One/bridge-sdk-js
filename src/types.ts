export enum Chain {
  Ethereum = "eth",
  Near = "near",
  Solana = "sol",
  Arbitrum = "arb",
  Base = "base",
}

export type OmniAddress = `${Chain}:${string}`

export enum Status {
  Pending = 0,
  Completed = 1,
  Failed = 2,
}

export interface TransferMessage {
  tokenAddress: OmniAddress
  amount: bigint
  fee: bigint
  nativeFee: bigint
  recipient: OmniAddress
  message: string | null
}

export interface OmniTransfer {
  txId: string
  nonce: bigint
  transferMessage: TransferMessage
}

export interface Fee {
  fee: bigint
  nativeFee: bigint
}
