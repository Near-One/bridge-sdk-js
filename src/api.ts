// api.ts
import { type ChainKind, type Fee, type OmniAddress, Status } from "./types"

export interface ApiTransferResponse {
  id: {
    origin_chain: "Eth" | "Near" | "Sol" | "Arb" | "Base"
    origin_nonce: number
  }
  status: "Initialized" | "FinalisedOnNear" | "Finalised"
  token: string
  amount: number
  recipient: string
  sender: string
  fee: {
    fee: number
    native_fee: number
  }
  msg: string
}

export interface ApiFeeResponse {
  native_token_fee: number
  transferred_token_fee: number | null
  usd_fee: number
}

export class OmniBridgeAPI {
  private baseUrl: string

  constructor(network: "testnet" | "mainnet") {
    this.baseUrl =
      network === "testnet"
        ? "https://testnet.api.bridge.nearone.org"
        : "https://api.bridge.nearone.org"
  }

  public getBaseUrl(): string {
    return this.baseUrl
  }

  async getTransferStatus(originChain: Chain, nonce: bigint): Promise<Status> {
    const params = new URLSearchParams({
      origin_chain: originChain,
      origin_nonce: nonce.toString(),
    })

    const response = await fetch(`${this.baseUrl}/api/v1/transfer?${params}`)
    if (!response.ok) {
      throw new Error(`API request failed: ${response.statusText}`)
    }

    const data: ApiTransferResponse = await response.json()

    switch (data.status) {
      case "Initialized":
        return Status.Pending
      case "FinalisedOnNear":
      case "Finalised":
        return Status.Completed
      default:
        return Status.Failed
    }
  }

  async getFee(sender: OmniAddress, recipient: OmniAddress, tokenAddress: string): Promise<Fee> {
    const params = new URLSearchParams({
      sender: sender,
      recipient: recipient,
      token: tokenAddress,
    })

    const response = await fetch(`${this.baseUrl}/api/v1/transfer-fee?${params}`)
    if (!response.ok) {
      throw new Error(`API request failed: ${response.statusText}`)
    }

    const data: ApiFeeResponse = await response.json()

    return {
      fee: data.transferred_token_fee ? BigInt(data.transferred_token_fee) : BigInt(0),
      nativeFee: BigInt(data.native_token_fee),
    }
  }
}
