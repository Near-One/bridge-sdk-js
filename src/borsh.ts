import { BorshSchema, borshSerialize } from "borsher"
import type { BindTokenArgs, FinDeployTokenArgs, FinTransferArgs, OmniAddress } from "./types"
import { Chain } from "./types"

const chainKindSchema = BorshSchema.Enum({
  Eth: BorshSchema.Unit,
  Near: BorshSchema.Unit,
  Sol: BorshSchema.Unit,
  Arb: BorshSchema.Unit,
  Base: BorshSchema.Unit,
})

const omniAddressSchema = BorshSchema.Enum({
  Eth: BorshSchema.String,
  Near: BorshSchema.String,
  Sol: BorshSchema.String,
  Arb: BorshSchema.String,
  Base: BorshSchema.String,
})

const transferIdSchema = BorshSchema.Struct({
  origin_chain: BorshSchema.String,
  origin_nonce: BorshSchema.u64,
})

const finTransferMessageSchema = BorshSchema.Struct({
  transfer_id: transferIdSchema,
  fee_recipient: BorshSchema.String,
  amount: BorshSchema.String,
  emitter_address: omniAddressSchema,
})

const logMetadataMessageSchema = BorshSchema.Struct({
  token_address: omniAddressSchema,
  name: BorshSchema.String,
  symbol: BorshSchema.String,
  decimals: BorshSchema.u8,
  emitter_address: omniAddressSchema,
})

const deployTokenMessageSchema = BorshSchema.Struct({
  token: BorshSchema.String,
  token_address: BorshSchema.String,
  emitter_address: omniAddressSchema,
})

const proverResultSchema = BorshSchema.Enum({
  InitTransfer: finTransferMessageSchema,
  FinTransfer: finTransferMessageSchema,
  DeployToken: deployTokenMessageSchema,
  LogMetadata: logMetadataMessageSchema,
})

const deployTokenArgsSchema = BorshSchema.Struct({
  chain_kind: chainKindSchema,
  prover_args: BorshSchema.Vec(BorshSchema.u8),
})

export type SerializedArgs<T> = Omit<T, "prover_args"> & {
  prover_args: Uint8Array
}

function serializeOmniAddress(addr: OmniAddress): { [key: string]: string } {
  const [chain, address] = addr.split(":")

  switch (chain) {
    case "eth":
      return { Eth: address }
    case "near":
      return { Near: address }
    case "sol":
      return { Sol: address }
    case "arb":
      return { Arb: address }
    case "base":
      return { Base: address }
    default:
      throw new Error(`Unknown chain: ${chain}`)
  }
}

function serializeChainKind(chain: Chain): { [key: string]: null } {
  switch (chain) {
    case Chain.Ethereum:
      return { Eth: null }
    case Chain.Near:
      return { Near: null }
    case Chain.Solana:
      return { Sol: null }
    case Chain.Arbitrum:
      return { Arb: null }
    case Chain.Base:
      return { Base: null }
  }
}

export function serializeDeployTokenArgs(args: FinDeployTokenArgs) {
  const emitter_address = serializeOmniAddress(args.prover_args.emitter_address)
  const token_address = serializeOmniAddress(args.prover_args.token_address)
  const wrappedMessage = {
    LogMetadata: {
      ...args.prover_args,
      token_address,
      emitter_address,
    },
  }
  const prover_args_vec: Uint8Array = borshSerialize(proverResultSchema, wrappedMessage)
  const deployTokenArgs = {
    chain_kind: serializeChainKind(args.chain_kind),
    prover_args: Array.from(prover_args_vec),
  }
  const result = borshSerialize(deployTokenArgsSchema, deployTokenArgs)
  return result
}

export function serializeBindTokenArgs(args: BindTokenArgs): SerializedArgs<BindTokenArgs> {
  const wrappedMessage = {
    DeployToken: args.prover_args,
  }
  return {
    ...args,
    prover_args: borshSerialize(proverResultSchema, wrappedMessage),
  }
}

export function serializeFinTransferArgs(args: FinTransferArgs): SerializedArgs<FinTransferArgs> {
  const wrappedMessage = {
    FinTransfer: args.prover_args,
  }
  return {
    ...args,
    prover_args: borshSerialize(proverResultSchema, wrappedMessage),
  }
}
