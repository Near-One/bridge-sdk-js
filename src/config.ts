export type NetworkType = "mainnet" | "testnet"

const ADDRESSES = {
  mainnet: {
    arb: "0xd025b38762B4A4E36F0Cde483b86CB13ea00D989",
    base: "0xd025b38762B4A4E36F0Cde483b86CB13ea00D989",
    bnb: "0x073C8a225c8Cf9d3f9157F5C1a1DbE02407f5720",
    eth: "0xe00c629afaccb0510995a2b95560e446a24c85b9",
    near: "omni.bridge.near",
    sol: {
      locker: "dahPEoZGXfyV58JqqH85okdHmpN8U2q8owgPUXSCPxe",
      wormhole: "worm2ZoG2kUd4vFXhvjh93UUH596ayRfgQ2MgjNMTth",
      shimProgram: "EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX",
      eventAuthority: "HQS31aApX3DDkuXgSpV9XyDUNtFgQ31pUn5BNWHG2PSp",
    },
    btc: {
      network: "mainnet" as const,
      apiUrl: "https://blockstream.info/api",
      mempoolUrl: "https://mempool.space/api",
      rpcUrl: "https://bitcoin-rpc.publicnode.com",
      btcConnector: "btc-connector.bridge.near",
      btcToken: "nbtc.bridge.near",
      bitcoinRelayer: "satoshi_optwo.near",
    },
    zcash: {
      network: "mainnet" as const,
      apiUrl: "https://zcash-mainnet.gateway.tatum.io/",
      rpcUrl: "https://zcash-mainnet.gateway.tatum.io/",
      zcashConnector: "zcash-connector.bridge.near",
      zcashToken: "nzec.bridge.near",
    },
  },
  testnet: {
    arb: "0x0C981337fFe39a555d3A40dbb32f21aD0eF33FFA",
    base: "0xa56b860017152cD296ad723E8409Abd6e5D86d4d",
    bnb: "0x7Fd1E9F9ed48ebb64476ba9E06e5F1a90e31DA74",
    eth: "0x68a86e0Ea5B1d39F385c1326e4d493526dFe4401",
    near: "omni.n-bridge.testnet",
    sol: {
      locker: "862HdJV59Vp83PbcubUnvuXc4EAXP8CDDs6LTxFpunTe",
      wormhole: "3u8hJUVTA4jH1wYAyUur7FFZVQ8H635K3tSHHF4ssjQ5",
      shimProgram: "EtZMZM22ViKMo4r5y4Anovs3wKQ2owUmDpjygnMMcdEX",
      eventAuthority: "HQS31aApX3DDkuXgSpV9XyDUNtFgQ31pUn5BNWHG2PSp",
    },
    btc: {
      network: "testnet" as const,
      apiUrl: "https://blockstream.info/testnet/api",
      mempoolUrl: "https://mempool.space/testnet/api",
      rpcUrl: "https://bitcoin-testnet-rpc.publicnode.com",
      btcConnector: "brg-dev.testnet",
      btcToken: "nbtc.n-bridge.testnet",
      bitcoinRelayer: "cosmosfirst.testnet",
    },
    zcash: {
      network: "testnet" as const,
      apiUrl: "https://zcash-testnet.gateway.tatum.io/",
      rpcUrl: "https://zcash-testnet.gateway.tatum.io/",
      zcashConnector: "zcash_connector.n-bridge.testnet",
      zcashToken: "nzcash.n-bridge.testnet",
    },
  },
} as const

let selectedNetwork: NetworkType = "mainnet"

export function setNetwork(network: NetworkType) {
  selectedNetwork = network
}

export function getNetwork(): NetworkType {
  return selectedNetwork
}

export const addresses = {
  get arb() {
    return ADDRESSES[selectedNetwork].arb
  },
  get base() {
    return ADDRESSES[selectedNetwork].base
  },
  get bnb() {
    return ADDRESSES[selectedNetwork].bnb
  },
  get eth() {
    return ADDRESSES[selectedNetwork].eth
  },
  get near() {
    return ADDRESSES[selectedNetwork].near
  },
  get sol() {
    return ADDRESSES[selectedNetwork].sol
  },
  get btc() {
    return ADDRESSES[selectedNetwork].btc
  },
  get zcash() {
    return ADDRESSES[selectedNetwork].zcash
  },
  get network() {
    return selectedNetwork
  },
}
