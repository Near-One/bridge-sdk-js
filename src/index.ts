// Export main client and factory

// Export API types and utilities
export * from "./api.js"
export * from "./client.js"
// Export specific clients
export * from "./clients/index.js"
export * from "./config.js"
export * from "./factory.js"
// Export proofs
export * from "./proofs/index.js"
// Export type definitions
export * from "./types/index.js"
export * from "./utils/index.js"
export * from "./utxo/index.js"
