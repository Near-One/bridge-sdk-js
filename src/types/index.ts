export * from "./chain.js"
export * from "./common.js"
export * from "./events.js"
export * from "./evm.js"
export * from "./locker.js"
export * from "./mpc.js"
export * from "./omni.js"
export * from "./prover.js"
export * from "./sol.js"
