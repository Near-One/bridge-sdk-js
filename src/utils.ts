import type { Chain, OmniAddress } from "./types"

// Helper function to construct OmniAddress
export const omniAddress = (chain: Chain, address: string): OmniAddress => {
  return `${chain}:${address}`
}

// Get chain from OmniAddress
export const getChain = (addr: OmniAddress): Chain => {
  const [chain] = addr.split(":") as [Chain]
  return chain
}
