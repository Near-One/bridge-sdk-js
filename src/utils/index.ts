export * from "./chain"
export * from "./decimals"
