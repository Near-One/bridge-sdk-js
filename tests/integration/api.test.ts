import { beforeEach, describe, expect, it } from "vitest"
import { OmniBridgeAPI } from "../../src/api"
import { setNetwork } from "../../src/config"
import { ChainKind, type OmniAddress } from "../../src/types"
import { omniAddress } from "../../src/utils"

describe("OmniBridgeAPI Integration Tests", () => {
  let api: OmniBridgeAPI

  beforeEach(() => {
    setNetwork("testnet")
    api = new OmniBridgeAPI()
  })

  describe("getFee", () => {
    it("should fetch real fee information", async () => {
      const sender: OmniAddress = omniAddress(ChainKind.Near, "bridge-sender.testnet")
      const recipient: OmniAddress = omniAddress(
        ChainKind.Eth,
        "0x000000F8637F1731D906643027c789EFA60BfE11",
      )
      const tokenAddress: OmniAddress = "near:warp.testnet"

      const fee = await api.getFee(sender, recipient, tokenAddress)

      // Check structure and types
      expect(fee).toEqual({
        native_token_fee: expect.toBeOneOf([expect.any(BigInt), null]),
        transferred_token_fee: expect.toBeOneOf([expect.any(BigInt), null]),
        usd_fee: expect.any(Number),
      })

      // Check valid ranges
      if (fee.native_token_fee !== null) {
        expect(fee.native_token_fee >= 0n).toBe(true)
      }
      if (fee.transferred_token_fee !== null) {
        expect(fee.transferred_token_fee >= 0n).toBe(true)
      }
      expect(fee.usd_fee >= 0).toBe(true)
    })

    it("should handle real API errors gracefully", async () => {
      const sender: OmniAddress = omniAddress(ChainKind.Eth, "invalid")
      const recipient: OmniAddress = omniAddress(ChainKind.Sol, "invalid")
      const tokenAddress: OmniAddress = "near:invalid"

      await expect(api.getFee(sender, recipient, tokenAddress)).rejects.toThrow()
    })
  })
  describe("getTransferStatus", () => {
    it("should fetch status for a known transfer", async () => {
      const status = await api.getTransferStatus("Eth", 53)
      expect(status).toMatchSnapshot()
    })

    it("should handle non-existent transfer", async () => {
      await expect(api.getTransferStatus("Eth", 999999999)).rejects.toThrow("Resource not found")
    })
  })
  describe("getTransfer", () => {
    it("should fetch transfer details for a known transfer", async () => {
      const transfer = await api.getTransfer("Eth", 53)

      // Need to convert BigInts to strings for snapshot
      const snapshotSafeTransfer = {
        ...transfer,
        transfer_message: {
          ...transfer.transfer_message,
          fee: {
            fee: transfer.transfer_message.fee.fee.toString(),
            native_fee: transfer.transfer_message.fee.native_fee.toString(),
          },
        },
      }

      expect(snapshotSafeTransfer).toMatchSnapshot()
    })

    it("should handle non-existent transfer", async () => {
      await expect(api.getTransfer("Eth", 999999999)).rejects.toThrow("Resource not found")
    })
  })
  describe("findOmniTransfers", () => {
    it("should fetch transfers for a specific transaction", async () => {
      const txId = "0x0b08b481f24e9df5fc5988777933796173e1f5ef9aa4878557df0a4f5b7d8ad0"
      const transfers = await api.findOmniTransfers({ transaction_id: txId })

      // Convert BigInts to strings for snapshot
      const snapshotSafeTransfers = transfers.map((transfer) => ({
        ...transfer,
        transfer_message: {
          ...transfer.transfer_message,
          fee: {
            fee: transfer.transfer_message.fee.fee.toString(),
            native_fee: transfer.transfer_message.fee.native_fee.toString(),
          },
        },
      }))

      expect(snapshotSafeTransfers).toMatchSnapshot()
    })

    it("should fetch transfers for a specific sender", async () => {
      const sender = "near:r-near.testnet"
      const transfers = await api.findOmniTransfers({ sender })

      // Verify structure and pagination
      expect(transfers).toBeInstanceOf(Array)
      expect(transfers.length).toBeLessThanOrEqual(10)

      if (transfers.length > 0) {
        // Check first transfer structure
        const transfer = transfers[0]
        expect(transfer).toEqual({
          id: {
            origin_chain: expect.stringMatching(/^(Eth|Near|Sol|Arb|Base)$/),
            origin_nonce: expect.any(Number),
          },
          initialized: expect.any(Object),
          finalised_on_near: expect.toBeOneOf([expect.anything(), undefined, null]), // null or transaction object
          finalised: expect.toBeOneOf([expect.anything(), undefined, null]), // null or transaction object
          transfer_message: {
            token: expect.any(String),
            amount: expect.any(Number),
            sender: expect.any(String),
            recipient: expect.any(String),
            fee: {
              fee: expect.any(BigInt),
              native_fee: expect.any(BigInt),
            },
            msg: expect.any(String),
          },
          updated_fee: expect.any(Array),
        })
      }
    })
  })
})
