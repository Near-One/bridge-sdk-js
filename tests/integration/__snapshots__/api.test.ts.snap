// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OmniBridgeAPI Integration Tests > findOmniTransfers > should fetch transfers for a specific transaction 1`] = `
[
  {
    "claimed": null,
    "fast_finalised": null,
    "fast_finalised_on_near": null,
    "finalised": null,
    "finalised_on_near": null,
    "id": {
      "origin_chain": "Near",
      "origin_nonce": 1,
    },
    "initialized": {
      "NearReceipt": {
        "block_height": 137141103,
        "block_timestamp_seconds": 1736635744,
        "transaction_hash": "DQM4d3B6Jxr4qnvGay4bpZ7aTQQogpUgQxVLAVWk5doF",
      },
    },
    "signed": null,
    "transfer_message": {
      "amount": 10000n,
      "fee": {
        "fee": 20n,
        "native_fee": 10n,
      },
      "msg": "",
      "recipient": "base:0xcf6462b9fce5af3e6c660c83453eca18ff468773",
      "sender": "near:frolik.near",
      "token": "near:wrap.near",
    },
    "updated_fee": [],
  },
]
`;

exports[`OmniBridgeAPI Integration Tests > getAllowlistedTokens > should return consistent data structure 1`] = `
{
  "1f9840a85d5af5bf1d1762f925bdaddc4201f984.factory.bridge.near": "eth:0x1f9840a85d5af5bf1d1762f925bdaddc4201f984",
  "2260fac5e5542a773aa44fbcfedf7c193bc2c599.factory.bridge.near": "eth:0x2260fac5e5542a773aa44fbcfedf7c193bc2c599",
  "6b175474e89094c44da98b954eedeac495271d0f.factory.bridge.near": "eth:0x6b175474e89094c44da98b954eedeac495271d0f",
  "a0b86991c6218b36c1d19d4a2e9eb0ce3606eb48.factory.bridge.near": "eth:0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
  "aaaaaa20d9e0e2461697782ef11675f668207961.factory.bridge.near": "eth:0xaaaaaa20d9e0e2461697782ef11675f668207961",
  "arb.omdep.near": "eth:0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
  "base.omdep.near": "eth:0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
  "eth.bridge.near": "eth:0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
  "sol.omdep.near": "sol:So11111111111111111111111111111111111111112",
  "token.publicailab.near": "near:token.publicailab.near",
  "token.sweat": "near:token.sweat",
  "wrap.near": "near:wrap.near",
}
`;

exports[`OmniBridgeAPI Integration Tests > getTransfer > should fetch transfer details for a known transfer 1`] = `
{
  "claimed": null,
  "fast_finalised": null,
  "fast_finalised_on_near": null,
  "finalised": {
    "EVMLog": {
      "block_height": 26239591,
      "block_timestamp_seconds": 1739268529,
      "transaction_hash": "0xf802fcc209d7d3bbb6b90393c06f3c821b34a9d1c3b2beb42cd1e7468ea1cfce",
    },
  },
  "finalised_on_near": null,
  "id": {
    "origin_chain": "Near",
    "origin_nonce": 22,
  },
  "initialized": {
    "NearReceipt": {
      "block_height": 139475752,
      "block_timestamp_seconds": 1739268494,
      "transaction_hash": "7wLmrkv9K85aWT8KKEn7roAyFwTT4FS3ZrdUMvxLkym8",
    },
  },
  "signed": null,
  "transfer_message": {
    "amount": 1000000000000000000000n,
    "fee": {
      "fee": 0n,
      "native_fee": 3483371056966369000000n,
    },
    "msg": "",
    "recipient": "base:0xd51c5283b8727206bf9be2b2db4e5673efaf519c",
    "sender": "near:marior.near",
    "token": "near:burn-1411.meme-cooking.near",
  },
  "updated_fee": [],
}
`;

exports[`OmniBridgeAPI Integration Tests > getTransferStatus > should fetch status for a known transfer 1`] = `"Initialized"`;
